# 统计与大数据分析软件第四次作业

## 第一题

冰淇淋小店：冰淇淋小店是一种特殊的餐馆。

编写一个名为`IceCreamStand`的类，让它继承你为完成第三次作业练习2或练习5而编写的`Restaurant`类。

这两个版本的`Restaurant`类都可以，挑选你更喜欢的那个即可。

添加一个名为`flavors`的属性，用于存储一个由各种口味的冰淇淋组成的列表。

编写一个显示这些冰淇淋的方法。创建一个`IceCreamStand`实例，并调用这个方法。


In [4]:
# 我们先提前导入之前的类（不知道如何两个notebook文件如何互相导入）
class Restaurant:
    # 初始化属性：名称和类型
    def __init__(self, restaurant_name, cuisine_type):
        self.restaurant_name = restaurant_name
        self.cuisine_type = cuisine_type
        self.number_served = 0

    # 定义方法：描述餐馆
    def describe_restaurant(self):
        print(f"{self.restaurant_name} serves wonderful {self.cuisine_type}.")

    # 定义方法：餐馆营业中
    def open_restaurant(self):
        print(f"{self.restaurant_name} is open. Come on in!")

    # 添加方法：设置就餐人数
    def set_number_served(self, number):
        self.number_served = number
        print(
            f"In {self.restaurant_name}, {self.number_served} people have been served."
        )

    # 添加方法：增加就餐人数
    def increment_number_served(self, number):
        self.number_served += number
        print(f"Now {self.number_served} people have been served.")

In [5]:
# 创建子类，实现冰淇凌店
class IceCreamStand(Restaurant):
    def __init__(self, restaurant_name, cuisine_type):
        super().__init__(restaurant_name, cuisine_type)
        self.flavors = []

    def add_flavor(self, flavor):
        self.flavors.append(flavor)

    def show_flavors(self):
        print("Available flavors:")
        for flavor in self.flavors:
            print("- " + flavor)

In [6]:
Mixuebingcheng = IceCreamStand("Mixuebingcheng", "ice cream")
Mixuebingcheng.describe_restaurant()
Mixuebingcheng.add_flavor("vanilla")
Mixuebingcheng.add_flavor("chocolate")
Mixuebingcheng.add_flavor("strawberry")
Mixuebingcheng.show_flavors()

Mixuebingcheng serves wonderful ice cream.
Available flavors:
- vanilla
- chocolate
- strawberry


## 第二题

管理员：管理员是一种特殊的用户。

编写一个名为`Admin`的类，让它继承你为完成第三次作业中练习4或练习6而编写的`User`类。

添加一个名为`privileges`的属性，用于存储一个由字符串（如"`can add post`"、"`can delete post`"、"`can ban user`"等）组成的列表。

编写一个名为`show_privileges()`的方法，它显示管理员的权限。创建一个`Admin`实例，并调用这个方法。


In [7]:
# 我们先提前导入之前的类
class User:
    def __init__(self, first_name, last_name, gender, age, nationality):
        self.first_name = first_name
        self.last_name = last_name
        self.gender = gender
        self.age = age
        self.nationality = nationality
        self.login_attempts = 0

    def describe_user(self):
        print(
            f"The person's name: {self.first_name} {self.last_name}, gender: {self.gender}, age: {self.age}, nationality: {self.nationality}."
        )

    def greet_user(self):
        print(f"Hello, {self.first_name} {self.last_name}! Nice to meet you!")

    # 定义方法：注册次数增加
    def increment_login_attempts(self):
        self.login_attempts += 1

    # 定义方法：重置登录次数
    def reset_login_attempts(self):
        self.login_attempts = 0

    # 定义方法：显示登录次数
    def show_login_attempts(self):
        print(f"The person has logined the number of {self.login_attempts}. ")

In [8]:
# 子类
class User_add(User):
    def __init__(self, first_name, last_name, age, sex, nationality):
        super().__init__(first_name, last_name, age, sex, nationality)
        self.privileges = []

    # 存储功能
    def add_privileges(self, privilege):
        self.privileges.append(privilege)

    # 显示功能
    def show_privileges(self):
        print(f"The person has the privileges of {self.privileges}.")

In [9]:
Mike = User_add("Mike", "Jordan", "18", "male", "American")
Mike.add_privileges("can add post")
Mike.add_privileges("can delete post")
Mike.add_privileges("can ban user")
Mike.show_privileges()

The person has the privileges of ['can add post', 'can delete post', 'can ban user'].


## 第三题

权限：编写一个名为`Privileges`的类，它只有一个属性——`privileges`，其中存储了练习2所说的字符串列表。

将方法`show_privileges()`移到这个类中。在`Admin`类中，将一个`Privileges`实例用作其属性。

创建一个`Admin`实例，并使用方法`show_privileges()`来显示其权限。


In [10]:
class Privileges:
    def __init__(self):
        self.privileges = []

    def add_privileges(self, privilege):
        self.privileges.append(privilege)

    def show_privileges(self):
        print(f"The person has the privileges of {self.privileges}.")

In [11]:
class Admin(User):
    def __init__(self, first_name, last_name, age, gender, nationality):
        super().__init__(first_name, last_name, age, gender, nationality)
        self.privileges = Privileges()

In [12]:
Jack = Admin("Jack", "Ma", "male", 20, "Chinese")
Jack.describe_user()
Jack.privileges.add_privileges("can add post")
Jack.privileges.add_privileges("can delete post")
Jack.privileges.add_privileges("can ban user")
Jack.privileges.show_privileges()

The person's name: Jack Ma, gender: male, age: 20, nationality: Chinese.
The person has the privileges of ['can add post', 'can delete post', 'can ban user'].


## 第四题

编写一个`while`循环，提示用户输入其名字。

用户输入其名字后，在屏幕上打印一句问候语，并将一条访问记录添加到文件`guest_book.txt`中。

确保这个文件中的每条记录都独占一行。 


In [13]:
# 以追加模式打开文件，这样每次写入内容时会添加到文件末尾
with open("guest_book.txt", "a") as file:
    while True:
        # 提示用户输入名字
        name = input("Please input your name ('quit' will break): ")
        # 检查用户是否输入 'quit'，如果是则退出循环
        if name.lower() == "quit":
            break
        # 打印问候语
        print(f"Hello, {name}, welcome!")
        # 将用户名字添加到文件中，并在末尾添加换行符
        file.write(name + "\n")

Hello, 12, welcome!
Hello, 34, welcome!


## 第五题

编写一个程序，提示用户输入他喜欢的数字，并使用`json.dump()`将这个数字存储到文件中。

再编写一个程序，从文件中读取这个值，并打印消息“`I know your favorite number! It’s _____.`”。


In [14]:
import json

# 提示用户输入喜欢的数字
favorite_number = input("请输入你喜欢的数字: ")

# 文件名
filename = "favorite_number.json"

try:
    # 以写入模式打开文件
    with open(filename, "w") as f:
        # 使用 json.dump() 把数字存储到文件中
        json.dump(favorite_number, f)
    print(f"你的喜欢的数字 {favorite_number} 已成功存储到 {filename} 中。")
except ValueError:
    print("你输入的不是一个有效的数字，请输入整数。")

你的喜欢的数字 23 已成功存储到 favorite_number.json 中。


In [15]:
import json

# 文件名
filename = "favorite_number.json"

try:
    # 以读取模式打开文件
    with open(filename) as f:
        # 使用 json.load() 从文件中读取数字
        favorite_number = json.load(f)
    print(f"I know your favorite number! It’s {favorite_number}.")
except FileNotFoundError:
    print(f"文件 {filename} 未找到，请先运行存储数字的程序。")

I know your favorite number! It’s 23.


## 第六题

将练习5中的两个程序合而为一。如果存储了用户喜欢的数字，就向用户显示它，

否则提示用户输入他喜欢的数字并将其存储到文件中。

运行这个程序两次，看看它是否像预期的那样工作。

In [16]:
import json

# 定义存储用户喜欢数字的文件名
filename = "favorite_numbers.json"

try:
    # 尝试以读取模式打开文件
    with open(filename) as f:
        # 从文件中加载用户喜欢的数字
        favorite_number = json.load(f)
    # 如果成功读取到数字，向用户显示该数字
    print(f"I know your favorite number! It’s {favorite_number}.")
except FileNotFoundError:
    # 如果文件不存在，提示用户输入喜欢的数字
    favorite_number = input("请输入你喜欢的数字: ")
    try:
        # 尝试将输入转换为整数
        favorite_number = int(favorite_number)
        # 以写入模式打开文件
        with open(filename, "w") as f:
            # 将数字存储到文件中
            json.dump(favorite_number, f)
        print(f"你的喜欢的数字 {favorite_number} 已成功存储到 {filename} 中。")
    except ValueError:
        # 如果输入不是有效的整数，提示用户输入有效的整数
        print("你输入的不是一个有效的数字，请输入整数。")

I know your favorite number! It’s 150.


两次输出结果如下：

你的喜欢的数字 150 已成功存储到 favorite_numbers.json 中。

I know your favorite number! It’s 150.